### PR TITLE
Posix

### PR DESCRIPTION
**Pull request by hlangeveld**
This is the rebased branch for POSIX support, with fixes for most of the questions raised by @rylnd.

All examples are tested with the common POSIX compliant shells:

* bash
* dash (most restricted)
* ksh93 

Speed of dash and ksh93 is roughly on par, while bash drags a little.

What's currently lacking is a set of guidelines for keeping within the POSIX restrictions.
(E.g., no local variables - use subshells instead.)

In part this also refactors the main code into separate functions, making it easier to invoke `shpec` recursively independent of the current shell.

Major POSIX restrictions addressed:

1. `echo` `-e` option is not a POSIX feature. Frankly, `echo` has changed so much between versions of unix, gnu, and linux distros, that we'd be better of eliminating it entirely, in favour of `printf`.

2. `[[ ... ]]` auto-quoting `test` equivalent is not supported by POSIX.  Replaced with `[ ... ]` and
careful quoting.

3. Replaced `(( ... ))` arithmetic.  In 99% of all cases, it can be replaced with `: $(( ... ))`.
It's less pretty, not pleasing to the eye, but it works in the common shells.
The potential one percent exception is for conditionals, like `if (( ... )); then ...` or 
`(( ... )) && ...`.  None of those were present in this code.
Note: assignments like `: $(( var=... ))` will persist in the current shell, just like you'd expect from
`(( var=... ))`.

4. Bash allows for redefining what the POSIX standard calls [Special built-in](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_14) utilities,
while dash and ksh93 only allow redefinition of normal built-in utilities.  One of the assertions in the default shpec tests tried to redefine the special built-in `exit`.  I chose `false` instead, which has no special meaning *in* the POSIX shell definition, except that POSIX allows a shell to redefine it as a built-in, which it is in most shells these days.

### CODE REVIEW

File: bin/shpec [hunk 1]
```diff
@@ -1,41 +1,51 @@
-#!/usr/bin/env bash
-VERSION=0.1.2
-examples=0
-failures=0
-test_indent=0
-red="\033[0;31m"
-green="\033[0;32m"
-norm="\033[0m"
+#
```
Comment (rylnd) [inline bin/shpec]: Please delete this line.

File: bin/shpec [hunk 2]
```diff
@@ -44,97 +54,141 @@ stub_command() {
...
-
-    time for file in $files; do
-      . "$file"
-    done
 
-    [ $failures -eq 0 ] && color=$green || color=$red
-    echo -e "${color}${examples} examples, ${failures} failures${norm}"
+shpec() {
+  (
+  VERSION=$(
```
Comment (rylnd) [inline bin/shpec]: I like this definition of `$VERSION`, but unfortunately it breaks our existing [release target](https://github.com/rylnd/shpec/blob/master/Makefile#L7). I would suggest we just revert this change.

File: bin/shpec [hunk 3]
```diff
@@ -44,97 +54,141 @@ stub_command() {
...
+  )
+  examples=0
+  failures=0
+  test_indent=0
+  red="\033[0;31m"
+  green="\033[0;32m"
+  norm="\033[0m"
+
+  SHPEC_ROOT=${SHPEC_ROOT:-$([ -d './shpec' ] && echo './shpec' || echo '.')}
+  case "$1" in
```
Comment (rylnd) [inline bin/shpec]: This `case` is written in a different style than the existing ones. I prefer the
```bash
( value )
  statement
;;
```
style, but the important thing is consistency.

File: bin/shpec [hunk 4]
```diff
@@ -44,97 +54,141 @@ stub_command() {
...
+  ( -v | --version ) echo "$VERSION"
+  ;;
+  ( * )
+      matcher_files=$(
```
Comment (rylnd) [inline bin/shpec]: Wrong indentation here (3 spaces).

**Comment by rylnd** [comment]
@hlangeveld thank you for this, it looks great. Once those minor comments are addressed, this is good to go.

**Comment by locochris** [comment]
@ryland @hlangeveld - perhaps its worth running a shell linter?
see:
 * https://github.com/hlangeveld/shpec/pull/2
 * https://travis-ci.org/hlangeveld/shpec/jobs/59327837
(although i'm hoping there's an easier way to set up this haskell/cabal dependency in travis - the uncached madness is currently taking > 5mins.)

File: bin/shpec (hunk 1 above)
Comment (hlangeveld) [inline bin/shpec]: On 04/20/15 17:20, Ryland Herrick wrote:
> In bin/shpec <https://github.com/rylnd/shpec/pull/58#discussion_r28697899>:
>
>> @@ -1,41 +1,51 @@
>> -#!/usr/bin/env bash
>> -VERSION=0.1.2
>> -examples=0
>> -failures=0
>> -test_indent=0
>> -red="\033[0;31m"
>> -green="\033[0;32m"
>> -norm="\033[0m"
>> +#
>
> Please delete this line.

I see now why I kept that line in there.  It's entirely non-functional,
but it's related to vim syntax highlighting.

With the  `#! /usr/bin/env bash`, vim automatically does rich syntax highlights.
Those disappear when I remove the entire line.

When I keep just the plain `#` as the first line, we get some very basic highlights.

There is a way to fix it *and* get rid of that line:

Rename the script to `bin/shpec.sh` and create a packager and/or installer that removes the suffix.

The `.sh` suffix gives me full syntax highlighting, even without that line.

**Comment by rylnd** [comment]
Looks good to me. Merging!

**Comment by rylnd** [comment]
Everyone, rebase your branches! :wink:
